controllers/yawol-controller/loadbalancerset/loadbalancermachine_status_controller.go

```diff
@@ -8,52 +8,92 @@ import (
-	"k8s.io/apimachinery/pkg/runtime"
-	"k8s.io/client-go/tools/record"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/ratelimiter"
 )
 
 // LoadBalancerMachineReconciler reconciles service Objects with type LoadBalancer
-type LoadBalancerMachineStatusReconciler struct {
+type LBMStatusReconciler struct {
 	client.Client
-	Log         logr.Logger
-	Scheme      *runtime.Scheme
-	Recorder    record.EventRecorder
 	WorkerCount int
 	RateLimiter ratelimiter.RateLimiter
+
+	DeletionGracePeriod time.Duration
 }
 
+const deletionMarkerCondition corev1.NodeConditionType = "MarkedForDeletion"
+
+// TODO: add more resources
+
 // +kubebuilder:rbac:groups=core,resources=services,verbs=get;list;watch
-func (r *LoadBalancerMachineStatusReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = r.Log.WithValues("LoadBalancerMachineStatus", req.NamespacedName)
+func (r *LBMStatusReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
 
-	var loadBalancerMachine yawolv1beta1.LoadBalancerMachine
-	if err := r.Client.Get(ctx, req.NamespacedName, &loadBalancerMachine); err != nil {
+	loadBalancerMachine := &yawolv1beta1.LoadBalancerMachine{}
+	if err := r.Client.Get(ctx, req.NamespacedName, loadBalancerMachine); err != nil {
 		if client.IgnoreNotFound(err) == nil {
 			return ctrl.Result{}, nil
 		}
-		r.Log.Info("error retrieving LoadBalancerMachine", "lbm", req)
+		log.Info("error retrieving LoadBalancerMachine", "lbm", req)
 		return ctrl.Result{}, err
 	}
 
-	if shouldBeDeleted, reason := shouldMachineBeDeleted(loadBalancerMachine); shouldBeDeleted {
-		if loadBalancerMachine.DeletionTimestamp != nil {
-			r.Log.Info("LoadBalancerMachine failed ReadyChecks and is already deleted", "lbm", req, "Reason:", reason)
-			return ctrl.Result{}, nil
+	if loadBalancerMachine.DeletionTimestamp != nil {
+		return ctrl.Result{}, nil
+	}
+
+	if ok, condition := getDeletionCondition(loadBalancerMachine); ok && condition.Status == corev1.ConditionTrue {
+		if err := r.handleMarkedMachine(ctx, log, loadBalancerMachine, condition); err != nil {
+			return ctrl.Result{}, err
 		}
-		if err := r.Client.Delete(ctx, &loadBalancerMachine); err != nil {
+		// TODO: do we want to requeue more preciesley?
+		return ctrl.Result{RequeueAfter: 15 * time.Second}, nil
+	}
+
+	if shouldBeDeleted, reason := shouldMachineBeDeleted(loadBalancerMachine); shouldBeDeleted {
+		if err := r.markForDeletion(ctx, loadBalancerMachine, reason.Error()); err != nil {
 			return ctrl.Result{}, err
 		}
-		r.Log.Info("LoadBalancerMachine failed ReadyChecks and got deleted", "lbm", req, "Reason:", reason)
+		log.Info("LoadBalancerMachine failed ReadyChecks and is marked for deletion", "Reason:", reason)
 		return ctrl.Result{}, nil
```

Comment:
We need to requeue here. Otherwise, we will never delete the machine unless some update event or restart happens.